### ISSUE TITLE
Is there any plan to support --target=wasm64-wasi ?

### ISSUE THREAD

**Issue by renxuqiang** [WebAssembly/wasi-sdk#212]
when --target=wasm64-wasi will be supported?

**Comment by vshymanskyy** [comment]
Duplicate of https://github.com/WebAssembly/wasi-sdk/issues/185

**Comment by sunfishcode** [comment]
Yes, everything in LLVM's wasm support and wasi-libc was written with wasm64 in mind, so it's just a matter of hooking up the pieces and making sure things actually work.

**Comment by AnatolyBug** [comment]
Nooby question: building a code which uses f64s to wasm32-wasi - Would that lead to errors/unexpected behaviour or is that fine?

**Comment by sunfishcode** [comment]
`f64` is perfectly fine with wasm32-wasi. `i64` is fine too.

**Comment by abrown** [comment]
`wasm64-wasi` in wasi-sdk could use some help: if anyone is interested in working on this feel free to pick this up.

**Comment by icppWorld** [comment]
@abrown 
How can I help? Very interested to help make it happen but don't know where to start.

**Comment by sunfishcode** [comment]
The guidance from the WASI Subgroup is that wasm64-wasi should now target WASI 0.2, so the path forward here now looks like:
 - Support for 64-bit pointers and lengths in the [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md). I'm optimistic this will happen fairly soon.
 - Implement the 64-bit Canonical ABI in wit-bindgen.
 - Support for WASI 0.2 in wasi-libc, which is underway (https://github.com/WebAssembly/wasi-libc/pull/471 and others).
 - Once the above are done, add a wasm64-wasi target to wasi-libc.

**Comment by lum1n0us** [comment]
> Support for 64-bit pointers and lengths in the [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md).

Could u share more about that? It seems there are no new comments in https://github.com/WebAssembly/component-model/issues/22 for monthes.

**Comment by sunfishcode** [comment]
The CanonicalABI change will be conceptually straightforward; things like the [size function](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size) need to be aware that a list has a pointer and a length, and so on.

https://github.com/bytecodealliance/wit-bindgen/pull/870 gets wit-bindgen a step closer to memory64 support. Beyond this, its [size/etc. functions](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wit-parser/src/sizealign.rs) also need to know that a list has a pointer and a length.

**Comment by YuriArthas** [comment]
Now, 2025....

**Comment by richgel999** [comment]
Any news on this? We could really use 64-bit. Our codec needs a lot of RAM to process larger images/textures.
https://github.com/BinomialLLC/basis_universal/

**Comment by alexcrichton** [comment]
The current status of wasm64 support is:

* The component model [does not specify how memory64 interacts with ABI just yet](https://github.com/WebAssembly/component-model/issues/22). This is not considered difficult to fix, it's just that no one's prioritized it yet. This blocks the `wasm64-wasip2` target, for example.
* There are no plans to release `wasm64-wasi{,-threads,p1,p1-threads}` targets at this time. Development of WASIp1 has ceased and adding 64-bit support would require redefining the existing ABI, so it's not "just change this thing there".

A number of various issues in various repositories have been closed in favor of https://github.com/WebAssembly/component-model/issues/22 over time so I'm going to do that here as well. That issue is the current blocker for starting to add wasm64 support in wasi-libc, and then consequently over here in wasi-sdk. Once https://github.com/WebAssembly/component-model/issues/22 makes some more progress new tracking issues can be made with up-to-date items.
